main.go

```diff
@@ -1,177 +1,510 @@
+	RancherCatalogName string `json:"rancher_catalog_name"`
+}
+
+// tagsByBranch struct
+type tagsByBranch struct {
+	branches map[string]branch
 }
 
 type TagsYaml struct {
 	Tags []string `yaml:"tags"`
 }
 
-func CheckImage( image string) bool {
-	imageParts := strings.Split(image, ":");
+// branch struct
+type branch struct {
+	versions map[string]version
+}
+
+// version struct
+type version struct {
+	builds map[int]*Tag
+}
 
-	//if there is a tag
-	if (len(imageParts) != 2) {
-		return false;
+// Tag struct
+type Tag struct {
+	Tag     string
+	Count   int
+	Owner   string
+	Project string
+	Branch  string
+	Version string
+	Build   int
+	SHA     string
+}
+
+func getTagsFromYaml(workspace plugin.Workspace) []string {
+	path := filepath.Join(workspace.Path, "docker/src/.droneTags.yml")
+	file, err := ioutil.ReadFile(path)
+
+	if err != nil {
+		fmt.Println("error opening .droneTags.yml file", err)
+		os.Exit(1)
 	}
 
-	tag := imageParts[1]; //get it
-	//get the parts expecting OWNER_REPO_BRANCH_VERSION_BUILD_COMMIT
-	//as the branch could have underscores it needs to be parsed last.
-	index := strings.Index(tag, "_");
-	if index == -1 {
-		return false;
+	var droneTags TagsYaml
+	yaml.Unmarshal(file, &droneTags)
+	return droneTags.Tags
+}
+
+func stringInSlice(str string, list []string) bool {
+	for _, v := range list {
+		if v == str {
+			return true
+		}
+	}
+	return false
+}
+
+func main() {
+	fmt.Println("starting drone-rancher-catalog...")
+
+	var catalog = catalog{}
+	plugin.Param("workspace", &catalog.workspace)
+	plugin.Param("repo", &catalog.repo)
+	plugin.Param("build", &catalog.build)
+	plugin.Param("vargs", &catalog.vargs)
+	plugin.MustParse()
+
+	if len(catalog.vargs.DockerRepo) == 0 {
+		fmt.Println("ERROR: docker_repo: Docker Registry Repo to read tags from, not specified")
+		os.Exit(1)
+	}
+	if len(catalog.vargs.DockerUsername) == 0 {
+		fmt.Println("ERROR: docker_username: Docker Registry Username not specified")
+		os.Exit(1)
+	}
+	if len(catalog.vargs.DockerPassword) == 0 {
+		fmt.Println("ERROR: docker_password: Docker Registry Password not specified")
+		os.Exit(1)
+	}
+	if len(catalog.vargs.CatalogRepo) == 0 {
+		fmt.Println("ERROR: catalog_repo: GitHub Catalog Repo not specified")
+		os.Exit(1)
+	}
+	if len(catalog.vargs.GitHubToken) == 0 {
+		fmt.Println("ERROR: github_token: GitHub User Token not specified")
+		os.Exit(1)
+	}
+	if len(catalog.vargs.CowpokeURL) == 0 {
+		fmt.Println("ERROR: cowpoke_url: cowpoke url not specified")
+		os.Exit(1)
+	}
+	if len(catalog.vargs.RancherCatalogName) == 0 {
+		fmt.Println("ERROR: rancher_catalog_name: catalog name in rancher is not specified")
+		os.Exit(1)
+	}
+	if len(catalog.vargs.DockerURL) == 0 {
+		catalog.vargs.DockerURL = "https://registry.hub.docker.com/"
+	}
+	if len(catalog.vargs.GitHubUser) == 0 {
+		catalog.vargs.GitHubUser = catalog.build.Author
+	}
+	if len(catalog.vargs.GitHubEmail) == 0 {
+		catalog.vargs.GitHubEmail = catalog.build.Email
```

Comment:
@EvanSnapp is there a way you can make a function for these checks? Since the check is pretty much the same, with a different value. This screams refactor.